File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/debug/DebugPanelActivity.java
package io.muun.apollo.presentation.ui.debug;

import io.muun.apollo.R;
import io.muun.apollo.data.external.HoustonConfig;
import io.muun.apollo.presentation.ui.base.BaseActivity;
import io.muun.apollo.presentation.ui.base.BaseView;
import io.muun.apollo.presentation.ui.view.MuunButton;

import android.content.Context;
import android.content.Intent;
import android.text.InputType;
import android.widget.EditText;
import android.widget.TextView;
import androidx.appcompat.app.AlertDialog;
import butterknife.BindView;
import butterknife.OnClick;
import com.google.android.material.switchmaterial.SwitchMaterial;
import rx.functions.Action1;

import javax.inject.Inject;
import javax.validation.constraints.NotNull;

public class DebugPanelActivity extends BaseActivity<DebugPanelPresenter> implements BaseView {

    @Inject
    HoustonConfig houstonConfig;

    @BindView(R.id.debug_text_server_address)
    TextView serverAddress;

    @BindView(R.id.debug_button_fund_wallet_onchain)
    MuunButton fundWalletOnChain;

    @BindView(R.id.debug_button_fund_wallet_offchain)
    MuunButton fundWalletOffChain;

    @BindView(R.id.debug_button_generate_blocks)
    MuunButton generateBlock;

    @BindView(R.id.debug_button_drop_last_tx)
    MuunButton dropLastTx;

    @BindView(R.id.debug_button_drop_tx)
    MuunButton dropTx;

    @BindView(R.id.debug_button_undrop_tx)
    MuunButton undropTx;

    @BindView(R.id.debug_switch_allow_multi_session)
    SwitchMaterial allowMultiSession;

    @BindView(R.id.debug_button_expire_all_other_sessions)
    MuunButton expireAllOtherSessions;

    /**
     * Creates an intent to launch this activity.
     */
    public static Intent getStartActivityIntent(@NotNull Context context) {
        return new Intent(context, DebugPanelActivity.class);
    }

    @Override
    protected void inject() {
        getComponent().inject(this);
    }

    @Override
    protected int getLayoutResource() {
        return R.layout.debug_activity;
    }

    @Override
    protected void initializeUi() {
        super.initializeUi();

        fundWalletOnChain.setOnClickListener(v -> presenter.fundThisWalletOnChain());
        fundWalletOffChain.setOnClickListener(v -> presenter.fundThisWalletOffChain());
        generateBlock.setOnClickListener(v -> presenter.generateBlock());
        dropLastTx.setOnClickListener(v -> presenter.dropLastTxFromMempool());

        dropTx.setOnClickListener(v -> handleTxIdInput(presenter::dropTx));
        undropTx.setOnClickListener(v -> handleTxIdInput(presenter::undropTx));

        allowMultiSession.setOnCheckedChangeListener(
                (buttonView, isChecked) -> presenter.toggleMultiSessions()
        );
        expireAllOtherSessions.setOnClickListener(v -> presenter.expireAllSessions());
    }

    private void handleTxIdInput(Action1<String> handler) {
        final AlertDialog.Builder builder = new AlertDialog.Builder(this);
        builder.setTitle("Title");

        final EditText input = new EditText(this);
        input.setInputType(InputType.TYPE_CLASS_TEXT);
        builder.setView(input);

        builder.setPositiveButton("OK", (dialog, which) -> {
            final String inputText = input.getText().toString();
            handler.call(inputText);
        });
        builder.setNegativeButton("Cancel", (dialog, which) -> dialog.cancel());

        builder.show();
    }

    @Override
    protected void onResume() {
        super.onResume();
        serverAddress.setText(houstonConfig.getUrl());
    }

    /**
     * Handle fetch replace operations button click.
     */
    @OnClick(R.id.debug_button_fetch_operations)
    public void syncOperationsButtonClicked() {
        presenter.fetchReplaceOperations();
    }

    /**
     * Handle sync phone contacts button click.
     */
    @OnClick(R.id.debug_button_upload_phone_contacts)
    public void syncPhoneContactsButtonClicked() {
        presenter.scanReplacePhoneContacts();
    }

    /**
     * Handle sync contacts button click.
     */
    @OnClick(R.id.debug_button_fetch_contacts)
    public void syncContactsButtonClicked() {
        presenter.fetchReplaceContacts();
    }

    /**
     * Handle sync real time data button click.
     */
    @OnClick(R.id.debug_button_sync_real_time_data)
    public void syncFeeButtonClicked() {
        presenter.syncRealTimeData();
    }

    /**
     * Handle sync external address indexes button click.
     */
    @OnClick(R.id.debug_button_sync_external_addresses_indexes)
    public void syncExternalAddressesIndexesButtonClicked() {
        presenter.syncExternalAddressesIndexes();
    }

    /**
     * Handle check integrity button click.
     */
    @OnClick(R.id.debug_button_integrity_check)
    public void checkIntegrity() {
        presenter.checkIntegrity();
    }

    /**
     * Handle update FCM token button click.
     */
    @OnClick(R.id.debug_button_force_fcm_token_update)
    public void updateFcmToken() {
        presenter.updateFcmToken();
    }
}

File: android/apolloui/src/main/java/io/muun/apollo/presentation/ui/debug/DebugPanelActivity.kt
package io.muun.apollo.presentation.ui.debug

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.text.InputType
import android.view.LayoutInflater
import android.widget.CompoundButton
import android.widget.EditText
import androidx.appcompat.app.AlertDialog
import androidx.core.content.ContextCompat
import androidx.viewbinding.ViewBinding
import io.muun.apollo.R
import io.muun.apollo.data.external.HoustonConfig
import io.muun.apollo.data.nfc.api.NfcSession
import io.muun.apollo.databinding.DebugActivityBinding
import io.muun.apollo.presentation.ui.activity.extension.MuunDialog
import io.muun.apollo.presentation.ui.activity.extension.NfcReaderModeExtension
import io.muun.apollo.presentation.ui.base.BaseActivity
import io.muun.apollo.presentation.ui.base.BaseView
import io.muun.apollo.presentation.ui.view.RichText
import rx.functions.Action1
import timber.log.Timber
import javax.inject.Inject

class DebugPanelActivity : BaseActivity<DebugPanelPresenter>(), BaseView {

    companion object {
        /**
         * Creates an intent to launch this activity.
         */
        @JvmStatic
        fun getStartActivityIntent(context: Context): Intent {
            return Intent(context, DebugPanelActivity::class.java)
        }

        private enum class NfcPairingState {
            NONE,
            PAIRING,
            UNPAIRING
        }
    }

    @Inject
    lateinit var nfcReaderModeExtension: NfcReaderModeExtension

    @Inject
    lateinit var houstonConfig: HoustonConfig

    private val binding: DebugActivityBinding
        get() = getBinding() as DebugActivityBinding

    override fun bindingInflater(): (LayoutInflater) -> ViewBinding {
        return DebugActivityBinding::inflate
    }

    private var pairingState: NfcPairingState = NfcPairingState.NONE

    override fun inject() {
        component.inject(this)
    }

    override fun setUpExtensions() {
        super.setUpExtensions()
        addExtension(nfcReaderModeExtension)
    }

    override fun getLayoutResource(): Int {
        return R.layout.debug_activity
    }

    override fun initializeUi() {
        super.initializeUi()

        with(binding) {
            debugButtonFundWalletOnchain.setOnClickListener { presenter.fundThisWalletOnChain() }
            debugButtonFundWalletOffchain.setOnClickListener { presenter.fundThisWalletOffChain() }
            debugButtonGenerateBlocks.setOnClickListener { presenter.generateBlock() }
            debugButtonDropLastTx.setOnClickListener { presenter.dropLastTxFromMempool() }

            debugButtonDropTx.setOnClickListener {
                handleTxIdInput { txId: String ->
                    presenter.dropTx(txId)
                }
            }
            debugButtonUndropTx.setOnClickListener {
                handleTxIdInput { txId: String ->
                    presenter.undropTx(txId)
                }
            }

            debugButtonPairSecurityCard.setOnClickListener {
                pairingState = NfcPairingState.PAIRING
                // Turn on NFC reader Mode
                enableReaderMode()

                // Show Simple Message Dialog - prompt card tap
                MuunDialog.Builder()
                    .title("Pair Security Card")
                    .message("Please tap your device with the security card to confirm Pairing.")
                    .onDismiss {
                        disableReaderMode()
                        pairingState = NfcPairingState.NONE
                    }
                    .build()
                    .let(this@DebugPanelActivity::showDialog)
            }

            debugButtonUnpairSecurityCard.setOnClickListener {
                pairingState = NfcPairingState.UNPAIRING
                // Turn on NFC reader Mode
                enableReaderMode()

                // Show Simple Message Dialog - prompt card tap
                MuunDialog.Builder()
                    .title("Unpair Security Card")
                    .message("Please tap your device with the security card to confirm Unpairing.")
                    .onDismiss {
                        disableReaderMode()
                        pairingState = NfcPairingState.NONE
                    }
                    .build()
                    .let(this@DebugPanelActivity::showDialog)
            }

            debugButtonEnterDiagnosticMode.setOnClickListener {
                presenter.enterDiagnosticMode()
            }

            debugSwitchAllowMultiSession.setOnCheckedChangeListener { _: CompoundButton, _: Boolean ->
                presenter.toggleMultiSessions()
            }
            debugButtonExpireAllOtherSessions.setOnClickListener { presenter.expireAllSessions() }

            debugButtonFetchOperations.setOnClickListener {
                presenter.fetchReplaceOperations()
            }

            debugButtonUploadPhoneContacts.setOnClickListener {
                presenter.scanReplacePhoneContacts()
            }

            debugButtonFetchContacts.setOnClickListener {
                presenter.fetchReplaceContacts()
            }

            debugButtonSyncRealTimeData.setOnClickListener {
                presenter.syncRealTimeData()
            }

            debugButtonSyncExternalAddressesIndexes.setOnClickListener {
                presenter.syncExternalAddressesIndexes()
            }

            debugButtonIntegrityCheck.setOnClickListener {
                presenter.checkIntegrity()
            }

            debugButtonForceFcmTokenUpdate.setOnClickListener {
                presenter.updateFcmToken()
            }
        }
    }

    private fun handleTxIdInput(handler: Action1<String>) {
        val builder = AlertDialog.Builder(this)
        builder.setTitle("Title")

        val input = EditText(this)
        input.inputType = InputType.TYPE_CLASS_TEXT
        builder.setView(input)

        builder.setPositiveButton("OK") { _: DialogInterface, _: Int ->
            val inputText = input.text.toString()
            handler.call(inputText)
        }
        builder.setNegativeButton("Cancel") { dialog: DialogInterface, _: Int ->
            dialog.cancel()
        }

        builder.show()
    }

    override fun onResume() {
        super.onResume()
        binding.debugTextServerAddress.text = houstonConfig.url
    }

    override fun onNewNfcSession(nfcSession: NfcSession) {
        when (pairingState) {
            NfcPairingState.PAIRING -> {

                // TODO do proper error handling
                val success = try {
                    presenter.setUpSecurityCard(nfcSession)
                    true
                } catch (e: Exception) {
                    Timber.e(e)
                    false
                }

                if (success) {
                    runOnUiThread {
                        binding.nfcCardPaired.text = RichText("TRUE")
                            .setForegroundColor(ContextCompat.getColor(this, R.color.green_dark))
                            .setBold()

                        MuunDialog.Builder()
                            .title("Pair Security Card")
                            .message("Success!")
                            .build()
                            .let(this::showDialog)
                    }

                } else {
                    runOnUiThread {
                        MuunDialog.Builder()
                            .title("Pair Security Card")
                            .message("Error! See Debug logs or dismiss and try again")
                            .build()
                            .let(this::showDialog)
                    }
                }
            }

            NfcPairingState.UNPAIRING -> {

                // TODO do proper error handling
                val success = try {
                    presenter.resetSecuritytCard(nfcSession)
                    true
                } catch (e: Exception) {
                    Timber.e(e)
                    false
                }

                dismissDialog()

                if (success) {
                    runOnUiThread {
                        binding.nfcCardPaired.text = RichText("FALSE")
                            .setForegroundColor(ContextCompat.getColor(this, R.color.red))
                            .setBold()

                        MuunDialog.Builder()
                            .title("Unpair Security Card")
                            .message("Success!")
                            .build()
                            .let(this::showDialog)
                    }
                } else {
                    runOnUiThread {
                        MuunDialog.Builder()
                            .title("Unpair Security Card")
                            .message("Error! See Debug logs or dismiss and try again")
                            .build()
                            .let(this::showDialog)
                    }
                }
            }

            else -> {
                // Ignore
                Timber.d("NFC: onTagDiscovered when neither pairing nor unpairing")
            }
        }
    }

    private fun enableReaderMode() {
        nfcReaderModeExtension.enableReaderMode()
    }

    private fun disableReaderMode() {
        nfcReaderModeExtension.disableReaderMode()
    }
}
